Use smart pointers with PageLoadStateObserverBase subclasses and instance variables

ddkilzer opened the pull request:
#### 946a761a0a7f9b63ceda7211a863527fa110883c
<pre>
Use smart pointers with PageLoadStateObserverBase subclasses and instance variables
<a href="https://bugs.webkit.org/show_bug.cgi?id=257938">https://bugs.webkit.org/show_bug.cgi?id=257938</a>
&lt;rdar://110485192&gt;

Reviewed by Alex Christensen.

PageLoadStateObserverBase (a.k.a. PageLoadState::Observer) is stored as
a raw pointer in PageLoadState::m_observers, but subclass
NavigationState inherits CanMakeWeakPtr&lt;&gt;.  Be consistent by making
PageLoadStateObserverBase inherit CanMakeWeakPtr&lt;&gt; and propagate
necessary changes.

While here, make Cocoa instance variables in these classes use
WeakObjCPtr&lt;&gt; instead of raw pointers, and implement convenience
functions that return RetainPtr&lt;&gt; objects to avoid autoreleasing these
instance variables repeatedly for method calls.

* Source/WebKit/UIProcess/API/Cocoa/PageLoadStateObserver.h:
- Store m_object using WeakObjCPtr&lt;&gt;, and use WeakObjCPtr&lt;&gt;::get() to
  avoid autoreleasing m_object on every method call.

* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
(WebKit::NavigationState):
- Change PageLoadState::Observer to public inheritance so
  CanMakeWeakPtr&lt;&gt; may be used from PageLoadState::Observer.
- Stop inheriting CanMakeWeakPtr&lt;&gt; from NavigationState.
(WebKit::NavigationState::fromWebPage):
- Must return a pointer because navigationStates() stores
  WeakPtr&lt;NavigationState&gt; values.
(WebKit::NavigationState::navigationDelegate const):
(WebKit::NavigationState::historyDelegate const):
- Add const modifier to methods.
(WebKit::NavigationState::webView const): Add.
- Store m_viewView as WeakObjCPtr&lt;WKWebView&gt;.

* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::navigationStates):
- Change from HashMap&lt;&gt; to WeakHashMap&lt;&gt;.
(WebKit::NavigationState::NavigationState):
- Use `webView` parameter instead of m_webView to avoid autoreleasing
  the object.
(WebKit::NavigationState::~NavigationState):
(WebKit::NavigationState::fromWebPage):
- Update to return a pointer instead of a reference.
(WebKit::NavigationState::navigationDelegate const):
(WebKit::NavigationState::historyDelegate const):
- Add const modifier to methods.
(WebKit::NavigationState::navigationGestureDidBegin):
(WebKit::NavigationState::navigationGestureWillEnd):
(WebKit::NavigationState::navigationGestureDidEnd):
(WebKit::NavigationState::willRecordNavigationSnapshot):
(WebKit::NavigationState::navigationGestureSnapshotWasRemoved):
(WebKit::NavigationState::didRequestPasswordForQuickLookDocument):
(WebKit::NavigationState::didStopRequestingPasswordForQuickLookDocument):
(WebKit::NavigationState::didFirstPaint):
(WebKit::NavigationState::NavigationClient::didChangeBackForwardList):
(WebKit::NavigationState::NavigationClient::willGoToBackForwardListItem):
(WebKit::NavigationState::NavigationClient::decidePolicyForNavigationAction):
(WebKit::NavigationState::NavigationClient::contentRuleListNotification):
(WebKit::NavigationState::NavigationClient::decidePolicyForNavigationResponse):
(WebKit::NavigationState::NavigationClient::didStartProvisionalNavigation):
(WebKit::NavigationState::NavigationClient::didStartProvisionalLoadForFrame):
(WebKit::NavigationState::NavigationClient::didReceiveServerRedirectForProvisionalNavigation):
(WebKit::NavigationState::NavigationClient::willPerformClientRedirect):
(WebKit::NavigationState::NavigationClient::didPerformClientRedirect):
(WebKit::NavigationState::NavigationClient::didCancelClientRedirect):
(WebKit::NavigationState::NavigationClient::didFailProvisionalNavigationWithError):
(WebKit::NavigationState::NavigationClient::didFailProvisionalLoadWithErrorForFrame):
(WebKit::NavigationState::NavigationClient::didCommitNavigation):
(WebKit::NavigationState::NavigationClient::didCommitLoadForFrame):
(WebKit::NavigationState::NavigationClient::didFinishDocumentLoad):
(WebKit::NavigationState::NavigationClient::didFinishNavigation):
(WebKit::NavigationState::NavigationClient::didFinishLoadForFrame):
(WebKit::NavigationState::NavigationClient::didFailLoadDueToNetworkConnectionIntegrity):
(WebKit::NavigationState::NavigationClient::didChangeLookalikeCharacters):
(WebKit::NavigationState::NavigationClient::didFailNavigationWithError):
(WebKit::NavigationState::NavigationClient::didFailLoadWithErrorForFrame):
(WebKit::NavigationState::NavigationClient::didSameDocumentNavigation):
(WebKit::NavigationState::NavigationClient::renderingProgressDidChange):
(WebKit::NavigationState::NavigationClient::didReceiveAuthenticationChallenge):
(WebKit::NavigationState::NavigationClient::shouldAllowLegacyTLS):
(WebKit::NavigationState::NavigationClient::didNegotiateModernTLS):
(WebKit::NavigationState::NavigationClient::processDidTerminate):
(WebKit::NavigationState::NavigationClient::processDidBecomeResponsive):
(WebKit::NavigationState::NavigationClient::processDidBecomeUnresponsive):
(WebKit::NavigationState::NavigationClient::webCryptoMasterKey):
(WebKit::NavigationState::NavigationClient::navigationActionDidBecomeDownload):
(WebKit::NavigationState::NavigationClient::navigationResponseDidBecomeDownload):
(WebKit::NavigationState::NavigationClient::contextMenuDidCreateDownload):
(WebKit::NavigationState::NavigationClient::didStartLoadForQuickLookDocumentInMainFrame):
(WebKit::NavigationState::NavigationClient::didFinishLoadForQuickLookDocumentInMainFrame):
(WebKit::NavigationState::NavigationClient::decidePolicyForSOAuthorizationLoad):
(WebKit::NavigationState::HistoryClient::didNavigateWithNavigationData):
(WebKit::NavigationState::HistoryClient::didPerformClientRedirect):
(WebKit::NavigationState::HistoryClient::didPerformServerRedirect):
(WebKit::NavigationState::HistoryClient::didUpdateHistoryTitle):
(WebKit::NavigationState::willChangeIsLoading):
(WebKit::NavigationState::didChangeIsLoading):
(WebKit::NavigationState::willChangeTitle):
(WebKit::NavigationState::didChangeTitle):
(WebKit::NavigationState::willChangeActiveURL):
(WebKit::NavigationState::didChangeActiveURL):
(WebKit::NavigationState::willChangeHasOnlySecureContent):
(WebKit::NavigationState::didChangeHasOnlySecureContent):
(WebKit::NavigationState::willChangeNegotiatedLegacyTLS):
(WebKit::NavigationState::didChangeNegotiatedLegacyTLS):
(WebKit::NavigationState::willChangeWasPrivateRelayed):
(WebKit::NavigationState::didChangeWasPrivateRelayed):
(WebKit::NavigationState::willChangeEstimatedProgress):
(WebKit::NavigationState::didChangeEstimatedProgress):
(WebKit::NavigationState::willChangeCanGoBack):
(WebKit::NavigationState::didChangeCanGoBack):
(WebKit::NavigationState::willChangeCanGoForward):
(WebKit::NavigationState::didChangeCanGoForward):
(WebKit::NavigationState::willChangeNetworkRequestsInProgress):
(WebKit::NavigationState::didChangeNetworkRequestsInProgress):
(WebKit::NavigationState::willChangeCertificateInfo):
(WebKit::NavigationState::didChangeCertificateInfo):
(WebKit::NavigationState::willChangeWebProcessIsResponsive):
(WebKit::NavigationState::didChangeWebProcessIsResponsive):
(WebKit::NavigationState::didSwapWebProcesses):
- Make use of navigationDelegate() convenience method instead of calling
  m_navigationDelegate.get().
- Make use of new webView() convenience method to avoid autoreleasing
  m_webView each time.  Add nil checks as needed.
- Remove unneeded RetainPtr&lt;&gt;::get() calls.
- Change C-style casts to static_cast&lt;&gt;() for consistency.

* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
(WebKit::PageClientImplCocoa::webView const): Add.

* Source/WebKit/UIProcess/PageLoadState.cpp:
(WebKit::PageLoadState::~PageLoadState):
(WebKit::PageLoadState::addObserver):
(WebKit::PageLoadState::removeObserver):
(WebKit::PageLoadState::callObserverCallback):
- Update to work with m_observers storing WeakPtr&lt;Observer&gt;.

* Source/WebKit/UIProcess/PageLoadState.h:
(WebKit::PageLoadStateObserverBase):
- Inherit from CanMakeWeakPtr&lt;&gt;.
- Change m_observers from Vector&lt;Observer*&gt; to WeakHashSet&lt;Observer&gt;.

* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(MESSAGE_CHECK):
- Update to use webView().
(WebKit::PageClientImpl::isViewVisible):
(WebKit::PageClientImpl::isViewInWindow):
(WebKit::PageClientImpl::decidePolicyForGeolocationPermissionRequest):
(WebKit::PageClientImpl::didStartProvisionalLoadForMainFrame):
(WebKit::PageClientImpl::didFailProvisionalLoadForMainFrame):
(WebKit::PageClientImpl::didCommitLoadForMainFrame):
(WebKit::PageClientImpl::minimumZoomScale const):
(WebKit::PageClientImpl::showSafeBrowsingWarning):
(WebKit::PageClientImpl::clearSafeBrowsingWarning):
(WebKit::PageClientImpl::clearSafeBrowsingWarningIfForMainFrameNavigation):
(WebKit::PageClientImpl::effectiveAppearanceIsDark const):
(WebKit::PageClientImpl::effectiveUserInterfaceLevelIsElevated const):
(WebKit::PageClientImpl::takeViewSnapshot):
(WebKit::PageClientImpl::couldNotRestorePageState):
(WebKit::PageClientImpl::restorePageState):
(WebKit::PageClientImpl::restorePageCenterAndScale):
(WebKit::PageClientImpl::closeFullScreenManager):
(WebKit::PageClientImpl::isFullScreen):
(WebKit::PageClientImpl::enterFullScreen):
(WebKit::PageClientImpl::exitFullScreen):
(WebKit::PageClientImpl::lockFullscreenOrientation):
(WebKit::PageClientImpl::unlockFullscreenOrientation):
(WebKit::PageClientImpl::beganEnterFullScreen):
(WebKit::PageClientImpl::beganExitFullScreen):
(WebKit::PageClientImpl::didFinishLoadingDataForCustomContentProvider):
(WebKit::PageClientImpl::mimeTypesWithCustomContentProviders):
(WebKit::PageClientImpl::navigationGestureDidBegin):
(WebKit::PageClientImpl::navigationGestureWillEnd):
(WebKit::PageClientImpl::navigationGestureDidEnd):
(WebKit::PageClientImpl::willRecordNavigationSnapshot):
(WebKit::PageClientImpl::didRemoveNavigationGestureSnapshot):
(WebKit::PageClientImpl::didFinishNavigation):
(WebKit::PageClientImpl::didFailNavigation):
(WebKit::PageClientImpl::didSameDocumentNavigationForMainFrame):
(WebKit::PageClientImpl::didChangeBackgroundColor):
(WebKit::PageClientImpl::videoControlsManagerDidChange):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::applicationDidEnterBackground):
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::pinnedStateWillChange):
(WebKit::PageClientImpl::pinnedStateDidChange):
(WebKit::PageClientImpl::drawPageBorderForPrinting):
(WebKit::PageClientImpl::didShowContextMenu):
(WebKit::PageClientImpl::didDismissContextMenu):
(WebKit::PageClientImpl::showShareSheet):
(WebKit::PageClientImpl::navigationGestureDidBegin):
(WebKit::PageClientImpl::navigationGestureWillEnd):
(WebKit::PageClientImpl::navigationGestureDidEnd):
(WebKit::PageClientImpl::willRecordNavigationSnapshot):
(WebKit::PageClientImpl::didRemoveNavigationGestureSnapshot):
(WebKit::PageClientImpl::handleControlledElementIDResponse):
- Make use of new webView() convenience method to avoid autoreleasing
  m_webView each time.  Add nil checks as needed.
- Add nullptr checks after calling NavigationState::fromWebPage() since
  it now returns a pointer instead of a reference.

Canonical link: <a href="https://commits.webkit.org/265120@main">https://commits.webkit.org/265120@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cdc0a4aacfcd619eedef340ff5b4fda9ebdc0547

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9838 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10086 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10331 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11490 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9560 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12070 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10036 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12494 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9992 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10800 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8341 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11893 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8107 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16290 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9205 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9074 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12370 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9573 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7789 "4 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8739 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12962 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1122 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9348 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
```diff
@@ -1404,127 +1408,129 @@ static _WKSOAuthorizationLoadPolicy wkSOAuthorizationLoadPolicy(SOAuthorizationL
     // Transfer our background assertion from the old process to the new one.
     if (m_networkActivity)
-        m_networkActivity = m_webView->_page->process().throttler().backgroundActivity("Page Load"_s).moveToUniquePtr();
+        m_networkActivity = webView()->_page->process().throttler().backgroundActivity("Page Load"_s).moveToUniquePtr();
```
Oops this needs a `nil` check first.

ddkilzer commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
```diff
@@ -76,7 +76,7 @@
 #import <wtf/cocoa/Entitlements.h>
 
-#define MESSAGE_CHECK(assertion) MESSAGE_CHECK_BASE(assertion, m_webView.get()->_page->process().connection())
+#define MESSAGE_CHECK(assertion) MESSAGE_CHECK_BASE(assertion, webView()->_page->process().connection())
```
Wasn't sure the best way to add a `nil` check here, or if it was even needed (if `m_webView` should never be `nil` when this is used).

achristensen07 replied:
#define MESSAGE_CHECK(assertion) if (auto webView = this->webView()) { MESSAGE_CHECK_BASE(...); } { else { ASSERT_NOT_REACHED(); }

ddkilzer commented on Source/WebKit/UIProcess/PageLoadState.cpp:
```diff
@@ -42,6 +42,11 @@ PageLoadState::PageLoadState(WebPageProxy& webPageProxy)
+#if ASSERT_ENABLED
+    m_observers.removeAllMatching([](auto& observer) {
+        return !observer;
+    });
```
We could also remove `nullptr` observers in `PageLoadState::callObserverCallback()`  Not sure if that will catch them all, though, to make the assertion below never fire.

achristensen07 replied:
Could we make m_observers a WeakHashSet?

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
```diff
@@ -107,26 +107,36 @@
+        // If the weak WKWebView wasn't alive here, this will leave entries
+        // in navigationStates() and PageLoadState::m_observers. In Debug
+        // builds, m_observers is cleaned up in the PageLoadState destructor.
+        bool didRemove = navigationStates().removeIf([this](auto& entry) {
```
Let's make navigationStates() return a WeakHashMap instead of this.

Commit:
Use smart pointers with PageLoadStateObserverBase subclasses and instance variables
https://bugs.webkit.org/show_bug.cgi?id=257938
<rdar://110485192>

Reviewed by Alex Christensen.

PageLoadStateObserverBase (a.k.a. PageLoadState::Observer) is stored as
a raw pointer in PageLoadState::m_observers, but subclass
NavigationState inherits CanMakeWeakPtr<>.  Be consistent by making
PageLoadStateObserverBase inherit CanMakeWeakPtr<> and propagate
necessary changes.

While here, make Cocoa instance variables in these classes use
WeakObjCPtr<> instead of raw pointers, and implement convenience
functions that return RetainPtr<> objects to avoid autoreleasing these
instance variables repeatedly for method calls.

* Source/WebKit/UIProcess/API/Cocoa/PageLoadStateObserver.h:
- Store m_object using WeakObjCPtr<>, and use WeakObjCPtr<>::get() to
  avoid autoreleasing m_object on every method call.

* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
(WebKit::NavigationState):
- Change PageLoadState::Observer to public inheritance so
  CanMakeWeakPtr<> may be used from PageLoadState::Observer.
- Stop inheriting CanMakeWeakPtr<> from NavigationState.
(WebKit::NavigationState::fromWebPage):
- Must return a pointer because navigationStates() stores
  WeakPtr<NavigationState> values.
(WebKit::NavigationState::navigationDelegate const):
(WebKit::NavigationState::historyDelegate const):
- Add const modifier to methods.
(WebKit::NavigationState::webView const): Add.
- Store m_viewView as WeakObjCPtr<WKWebView>.

* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::navigationStates):
- Change from HashMap<> to WeakHashMap<>.
(WebKit::NavigationState::NavigationState):
- Use `webView` parameter instead of m_webView to avoid autoreleasing
  the object.
(WebKit::NavigationState::~NavigationState):
(WebKit::NavigationState::fromWebPage):
- Update to return a pointer instead of a reference.
(WebKit::NavigationState::navigationDelegate const):
(WebKit::NavigationState::historyDelegate const):
- Add const modifier to methods.
(WebKit::NavigationState::navigationGestureDidBegin):
(WebKit::NavigationState::navigationGestureWillEnd):
(WebKit::NavigationState::navigationGestureDidEnd):
(WebKit::NavigationState::willRecordNavigationSnapshot):
(WebKit::NavigationState::navigationGestureSnapshotWasRemoved):
(WebKit::NavigationState::didRequestPasswordForQuickLookDocument):
(WebKit::NavigationState::didStopRequestingPasswordForQuickLookDocument):
(WebKit::NavigationState::didFirstPaint):
(WebKit::NavigationState::NavigationClient::didChangeBackForwardList):
(WebKit::NavigationState::NavigationClient::willGoToBackForwardListItem):
(WebKit::NavigationState::NavigationClient::decidePolicyForNavigationAction):
(WebKit::NavigationState::NavigationClient::contentRuleListNotification):
(WebKit::NavigationState::NavigationClient::decidePolicyForNavigationResponse):
(WebKit::NavigationState::NavigationClient::didStartProvisionalNavigation):
(WebKit::NavigationState::NavigationClient::didStartProvisionalLoadForFrame):
(WebKit::NavigationState::NavigationClient::didReceiveServerRedirectForProvisionalNavigation):
(WebKit::NavigationState::NavigationClient::willPerformClientRedirect):
(WebKit::NavigationState::NavigationClient::didPerformClientRedirect):
(WebKit::NavigationState::NavigationClient::didCancelClientRedirect):
(WebKit::NavigationState::NavigationClient::didFailProvisionalNavigationWithError):
(WebKit::NavigationState::NavigationClient::didFailProvisionalLoadWithErrorForFrame):
(WebKit::NavigationState::NavigationClient::didCommitNavigation):
(WebKit::NavigationState::NavigationClient::didCommitLoadForFrame):
(WebKit::NavigationState::NavigationClient::didFinishDocumentLoad):
(WebKit::NavigationState::NavigationClient::didFinishNavigation):
(WebKit::NavigationState::NavigationClient::didFinishLoadForFrame):
(WebKit::NavigationState::NavigationClient::didFailLoadDueToNetworkConnectionIntegrity):
(WebKit::NavigationState::NavigationClient::didChangeLookalikeCharacters):
(WebKit::NavigationState::NavigationClient::didFailNavigationWithError):
(WebKit::NavigationState::NavigationClient::didFailLoadWithErrorForFrame):
(WebKit::NavigationState::NavigationClient::didSameDocumentNavigation):
(WebKit::NavigationState::NavigationClient::renderingProgressDidChange):
(WebKit::NavigationState::NavigationClient::didReceiveAuthenticationChallenge):
(WebKit::NavigationState::NavigationClient::shouldAllowLegacyTLS):
(WebKit::NavigationState::NavigationClient::didNegotiateModernTLS):
(WebKit::NavigationState::NavigationClient::processDidTerminate):
(WebKit::NavigationState::NavigationClient::processDidBecomeResponsive):
(WebKit::NavigationState::NavigationClient::processDidBecomeUnresponsive):
(WebKit::NavigationState::NavigationClient::webCryptoMasterKey):
(WebKit::NavigationState::NavigationClient::navigationActionDidBecomeDownload):
(WebKit::NavigationState::NavigationClient::navigationResponseDidBecomeDownload):
(WebKit::NavigationState::NavigationClient::contextMenuDidCreateDownload):
(WebKit::NavigationState::NavigationClient::didStartLoadForQuickLookDocumentInMainFrame):
(WebKit::NavigationState::NavigationClient::didFinishLoadForQuickLookDocumentInMainFrame):
(WebKit::NavigationState::NavigationClient::decidePolicyForSOAuthorizationLoad):
(WebKit::NavigationState::HistoryClient::didNavigateWithNavigationData):
(WebKit::NavigationState::HistoryClient::didPerformClientRedirect):
(WebKit::NavigationState::HistoryClient::didPerformServerRedirect):
(WebKit::NavigationState::HistoryClient::didUpdateHistoryTitle):
(WebKit::NavigationState::willChangeIsLoading):
(WebKit::NavigationState::didChangeIsLoading):
(WebKit::NavigationState::willChangeTitle):
(WebKit::NavigationState::didChangeTitle):
(WebKit::NavigationState::willChangeActiveURL):
(WebKit::NavigationState::didChangeActiveURL):
(WebKit::NavigationState::willChangeHasOnlySecureContent):
(WebKit::NavigationState::didChangeHasOnlySecureContent):
(WebKit::NavigationState::willChangeNegotiatedLegacyTLS):
(WebKit::NavigationState::didChangeNegotiatedLegacyTLS):
(WebKit::NavigationState::willChangeWasPrivateRelayed):
(WebKit::NavigationState::didChangeWasPrivateRelayed):
(WebKit::NavigationState::willChangeEstimatedProgress):
(WebKit::NavigationState::didChangeEstimatedProgress):
(WebKit::NavigationState::willChangeCanGoBack):
(WebKit::NavigationState::didChangeCanGoBack):
(WebKit::NavigationState::willChangeCanGoForward):
(WebKit::NavigationState::didChangeCanGoForward):
(WebKit::NavigationState::willChangeNetworkRequestsInProgress):
(WebKit::NavigationState::didChangeNetworkRequestsInProgress):
(WebKit::NavigationState::willChangeCertificateInfo):
(WebKit::NavigationState::didChangeCertificateInfo):
(WebKit::NavigationState::willChangeWebProcessIsResponsive):
(WebKit::NavigationState::didChangeWebProcessIsResponsive):
(WebKit::NavigationState::didSwapWebProcesses):
- Make use of navigationDelegate() convenience method instead of calling
  m_navigationDelegate.get().
- Make use of new webView() convenience method to avoid autoreleasing
  m_webView each time.  Add nil checks as needed.
- Remove unneeded RetainPtr<>::get() calls.
- Change C-style casts to static_cast<>() for consistency.

* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
(WebKit::PageClientImplCocoa::webView const): Add.

* Source/WebKit/UIProcess/PageLoadState.cpp:
(WebKit::PageLoadState::~PageLoadState):
(WebKit::PageLoadState::addObserver):
(WebKit::PageLoadState::removeObserver):
(WebKit::PageLoadState::callObserverCallback):
- Update to work with m_observers storing WeakPtr<Observer>.

* Source/WebKit/UIProcess/PageLoadState.h:
(WebKit::PageLoadStateObserverBase):
- Inherit from CanMakeWeakPtr<>.
- Change m_observers from Vector<Observer*> to WeakHashSet<Observer>.

* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(MESSAGE_CHECK):
- Update to use webView().
(WebKit::PageClientImpl::isViewVisible):
(WebKit::PageClientImpl::isViewInWindow):
(WebKit::PageClientImpl::decidePolicyForGeolocationPermissionRequest):
(WebKit::PageClientImpl::didStartProvisionalLoadForMainFrame):
(WebKit::PageClientImpl::didFailProvisionalLoadForMainFrame):
(WebKit::PageClientImpl::didCommitLoadForMainFrame):
(WebKit::PageClientImpl::minimumZoomScale const):
(WebKit::PageClientImpl::showSafeBrowsingWarning):
(WebKit::PageClientImpl::clearSafeBrowsingWarning):
(WebKit::PageClientImpl::clearSafeBrowsingWarningIfForMainFrameNavigation):
(WebKit::PageClientImpl::effectiveAppearanceIsDark const):
(WebKit::PageClientImpl::effectiveUserInterfaceLevelIsElevated const):
(WebKit::PageClientImpl::takeViewSnapshot):
(WebKit::PageClientImpl::couldNotRestorePageState):
(WebKit::PageClientImpl::restorePageState):
(WebKit::PageClientImpl::restorePageCenterAndScale):
(WebKit::PageClientImpl::closeFullScreenManager):
(WebKit::PageClientImpl::isFullScreen):
(WebKit::PageClientImpl::enterFullScreen):
(WebKit::PageClientImpl::exitFullScreen):
(WebKit::PageClientImpl::lockFullscreenOrientation):
(WebKit::PageClientImpl::unlockFullscreenOrientation):
(WebKit::PageClientImpl::beganEnterFullScreen):
(WebKit::PageClientImpl::beganExitFullScreen):
(WebKit::PageClientImpl::didFinishLoadingDataForCustomContentProvider):
(WebKit::PageClientImpl::mimeTypesWithCustomContentProviders):
(WebKit::PageClientImpl::navigationGestureDidBegin):
(WebKit::PageClientImpl::navigationGestureWillEnd):
(WebKit::PageClientImpl::navigationGestureDidEnd):
(WebKit::PageClientImpl::willRecordNavigationSnapshot):
(WebKit::PageClientImpl::didRemoveNavigationGestureSnapshot):
(WebKit::PageClientImpl::didFinishNavigation):
(WebKit::PageClientImpl::didFailNavigation):
(WebKit::PageClientImpl::didSameDocumentNavigationForMainFrame):
(WebKit::PageClientImpl::didChangeBackgroundColor):
(WebKit::PageClientImpl::videoControlsManagerDidChange):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::applicationDidEnterBackground):
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::pinnedStateWillChange):
(WebKit::PageClientImpl::pinnedStateDidChange):
(WebKit::PageClientImpl::drawPageBorderForPrinting):
(WebKit::PageClientImpl::didShowContextMenu):
(WebKit::PageClientImpl::didDismissContextMenu):
(WebKit::PageClientImpl::showShareSheet):
(WebKit::PageClientImpl::navigationGestureDidBegin):
(WebKit::PageClientImpl::navigationGestureWillEnd):
(WebKit::PageClientImpl::navigationGestureDidEnd):
(WebKit::PageClientImpl::willRecordNavigationSnapshot):
(WebKit::PageClientImpl::didRemoveNavigationGestureSnapshot):
(WebKit::PageClientImpl::handleControlledElementIDResponse):
- Make use of new webView() convenience method to avoid autoreleasing
  m_webView each time.  Add nil checks as needed.
- Add nullptr checks after calling NavigationState::fromWebPage() since
  it now returns a pointer instead of a reference.

Canonical link: https://commits.webkit.org/265120@main
